File: algorithms/code/leetcode/lc001_two_sum/lc001_two_sum.py
# https://leetcode.com/problems/two-sum/

import unittest


class Solution:
    def twoSum(self, nums: list[int], target: int) -> list[int]:
        mapped: dict[int, int] = {}
        for i, num in enumerate(nums):
            expected = target - num
            if expected in mapped:
                found = mapped[expected]
                return [found, i]
            mapped[num] = i
        return []


class TestCode(unittest.TestCase):
    def test_00(self) -> None:
        self.assertListEqual(Solution().twoSum([], 0), [])

    def test_1(self) -> None:
        self.assertListEqual(Solution().twoSum([2, 7, 11, 15], 9), [0, 1])

    def test_2(self) -> None:
        self.assertListEqual(Solution().twoSum([3, 2, 4], 6), [1, 2])

    def test_3(self) -> None:
        self.assertListEqual(Solution().twoSum([3, 3], 6), [0, 1])

File: algorithms/code/leetcode/lc001_two_sum/lc001_two_sum.rs
// https://leetcode.com/problems/two-sum/

use std::collections::HashMap;

pub struct Solution;

impl Solution {
    pub fn two_sum(nums: Vec<i32>, target: i32) -> Vec<i32> {
        let mut mapped: HashMap<i32, i32> = HashMap::new();
        for (i, &num) in nums.iter().enumerate() {
            let expected = target - num;
            if let Some(found) = mapped.get(&expected) {
                return vec![*found, i as i32];
            }
            mapped.insert(num, i as i32);
        }
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_00() {
        assert_eq!(Solution::two_sum(vec![], 0), []);
    }

    #[test]
    fn test_1() {
        assert_eq!(Solution::two_sum(vec![2, 7, 11, 15], 9), vec![0, 1]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::two_sum(vec![3, 2, 4], 6), vec![1, 2]);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::two_sum(vec![3, 3], 6), vec![0, 1]);
    }
}
